File: SelfProduct.java
//Using Running Product

class Solution {
    public int[] productExceptSelf(int[] nums) {
        int[] left = new int[nums.length];
        
        if(nums == null || nums.length == 0)
            return left;
        
        int rp = 1;
        left[0] = 1;
        
        //create the left array
        for(int i=1; i<nums.length; i++){
            rp = rp * nums[i-1];
            left[i] = rp;
        }
        
        //reset rp = 1 and compute the right side
        rp = 1;
        for(int i=nums.length-2; i>=0; i--){
            rp = rp * nums[i+1];
            left[i] = left[i] * rp;
        }
        return left;
    }
}

//Time Complexity :O(n)
//Space Complexity :O(1)

File: self_product.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        product =1
        zero_index = []
        for i in range(len(nums)):
            if nums[i] != 0:
                product = product*nums[i]
            else :
                zero_index.append(i)
        
        if len(zero_index)>1:
            return [0]*len(nums)
        elif len(zero_index)==1:
            temp = [0]*len(nums)
            temp[zero_index[0]] = product
            return temp
            
        else :
            for i in range(len(nums)):
                nums[i] = product//nums[i]
            
        return nums

        #time complexity  O(n) [iterate array twice]
        #space complexity O(n) [ zero index array]